### DISCUSSION TITLE
Individual page state

### DISCUSSION THREAD

**Discussion by cyz1901** [armanbilge/calico#330] [General]
Hey @armanbilge. I am currently trying to add backend support(http) to the page in a project using Calico.

So I chose the [https-dom](https://github.com/http4s/http4s-dom) library to integrate.

What has been confusing me all along is whether the calico library only supports passing in at the top.

Like in this [example](https://github.com/kubukoz/calico-smithy4s-demo/blob/main/front/shared/src/main/scala/com/kubukoz/lame/Client.scala)

If I only need to call the backend interface when a certain page is opened, or if I want to store certain states on a specific page (only applicable to that page), I don't want to persist them to the browser, only in memory of the current page.

My current idea is to put all the states at the entrance and then upload them to the required page, as we [discussed](https://github.com/cyz1901/WisdomGu/blob/6ee86e8046630bb44726ec9bc1133f69e29f2b77/modules/frontend/src/main/scala/per/cyz/wisdomGu/fronted/App.scala#L34).

I thought about it for a moment, and there will be two questions

1. If there are many things that need to be passed in, then many layers will be nested (this can be solved by splitting out more functions)

2. All states are initialized when the webpage is first opened, and as more pages expand, more and more data is stored in memory. When I switch to another page, because it is a globally transmitted item, will it cause the application to lag (imagine that a page will store a large amount of content about files and then store them in the `GlobalStore`. Then I switched pages, but these file data should still be stored in memory, right)

Perhaps in functional programming, I don't need to worry that everything has already been handled...:flushed:

Can you give me some advice if you have time?

**Comment by cyz1901** [comment]
And how to perform some side effects, such as network IO, on a defined `Resource [IO, HtmlElement[IO]` before initializing HtmlElement?

In my superficial understanding
```scala
   div("a", cls := "w-full h-full bg-slate-400 ml-[77px]").evalTap(_ =>
      IO {
        println("hello")
      }
    )
```

` println("hello")` Execute after element initialization is completed.

**Reply by armanbilge** [reply]
```scala
IO.println("hello").toResource *> div("a", cls := "w-full h-full bg-slate-400 ml-[77px]")
```

**Comment by armanbilge** [comment]
@cyz1901 have you seen this example? http://armanbilge.github.io/calico/demos/http4s.html

**Reply by cyz1901** [reply]
oh This is my negligence. :dizzy_face:

**Comment by armanbilge** [comment]
> What has been confusing me all along is whether the calico library only supports passing in at the top.

Every page can have its own state. For example:

```scala
def page1 = SignallingRef[IO].of(0).toResource.flatMap { counter =>
  // build page one here
}

def page2 = SignallingRef[IO].of(false).toResource.flatMap { toggle =>
  // build page two here
}
```

Let me know if that makes sense :)

**Reply by cyz1901** [reply]
Sorry, I just had time to try it out. The code works successfully. Thank you for your help. I think I have reached the threshold for using this framework. It is simple and intuitive.
